README.md
# SLOP
「 Sample Limning Operator Package 」

# Usage:
python3 slop-cli.py (or: py slop-cli.py)

# Pre-requisites:
python 3 on the system

pip on the system

following libraries installed through pip:
- pydub
- scipy

# Get in touch
twi: @elysian_tunes

email: analogpatch@gmail.com


files/changelog.txt
--- SLOP: CHANGELOG ---

2020 MAR 16: initial rollout of the thing. Basic functions, probably lots of bugs and failure points.

2020 MAR 17: new features: per-chunk amplitude normalisation and chunk length randomisation. menu slightly remade. some bugs fixed.

slop-cli.py

                #========           SLOP COMMAND LINE INTERFACE           ========#
                #======== PART OF SLOP (SAMPLE-LIMNING OPERATORS PACKAGE) ========#
                #========              V1.0    16 MAR 2020                ========#

'''
MAIN FUNCTION: - 
MAIN PURPOSE:  CLI version of SLOP
NOTES:         .
               .
               made by: Astro (aka Astro The Fox, Elysian Tunes, SYNTHFOX, etc..)
               original algorythm envision: Nyn Famitory
               special thanks to people who made libraries listed below there in [# imports]
               .
               visit:
                   https://elysiantunes.bandcamp.com      ~ my music
                   https://sfcs.neocities.org             ~ my DIY synthesizers
                   https://astrossoundhell.neocities.org  ~ my personal webpage
                   https://github.com/kouyouelysian       ~ my github profile
				   https://famitory.bandcamp.com/         ~ Famitory's music
			   .
'''

#--------------------------------------------------------------------------------------------------------
# imports

from inc import bitepacker
from inc import sampleslicer
from inc import matchengine
from inc import fftanalyzer
from inc import rangegenerator
from inc import packmanager

from os.path import exists, isfile, join, dirname, basename
import os
from sys import platform


#--------------------------------------------------------------------------------------------------------
# check if a menu option is one letter
def checkMenuOption(arg):
	if (arg.strip() == ''):
		return False
	if (len(arg)>1):
		return False
	return arg.upper()


#--------------------------------------------------------------------------------------------------------
# return false if no such dir
def checkDir(arg):
	if (exists(arg)):
		return arg
	return False


#--------------------------------------------------------------------------------------------------------
# return false if no such file
def checkFile(arg):
	if (os.path.isfile(arg)):
		return arg
	return False 


#--------------------------------------------------------------------------------------------------------
# put a postfix between file's name and extension
def postfixFileName(name, postfix):
	parts = os.path.basename(name).split('.')
	return parts[0] + postfix + "." + parts[1] 
 

#--------------------------------------------------------------------------------------------------------
# clear the console screen
def consoleClear():
	if platform == "windows":
		os.system('cls')
	else:
		os.system('clear')



#--------------------------------------------------------------------------------------------------------
# debug check a var
def check(arg):
	print(arg)
	input("\nany key to run next..")


#--------------------------------------------------------------------------------------------------------
# local vars

defaultBanner = "welcome to SLOP (Sample Limning Operator Package)"
consoleBanner = defaultBanner
badMessage = ""

defaultPackSettingsLocation = "./bitepack/packs.setting"
packPlaces = ["./bitepack/"]

aboutLocation = "./files/about.txt"
changelogLocation = "./files/changelog.txt"

#--------------------------------------------------------------------------------------------------------
# pre-runtime stuff


packSettings = packmanager.importSettings(defaultPackSettingsLocation)

for packTuple in packSettings["packsAvailable"]:
	if (not exists(packTuple["path"])):
		print("AAAA")
		input()
		packSettings = packmanager.rescan(packPlaces)
		break



#--------------------------------------------------------------------------------------------------------
# cli loop

while (1):

	# prepare
	consoleClear()
	print("\n" + consoleBanner + "\n")
	if (badMessage):
		badMessage = False
		consoleBanner = defaultBanner

	# print opts
	print("\n---- RUN:")
	print("M: MatchEngine")
	print("P: PackManager")
	print("B: BitePacker")
	print("\n---- READ:")
	print("A: about SLOP")
	print("C: changelog")
	print("\n---- OTHER:")
	print("X: exit software")
	opt = checkMenuOption(input("\n#: "))

	# process input option
	if (opt):

		# run bitepacker
		if (opt == "B"):
			consoleClear()
			print("\n" + "SLOP BitePacker" + "\n")
			print("\n" + "What to use BitePacker on?")

			print("\nS: single file")
			print("F: contents of a folder")
			print("L: list of files")
			print("D: done, go back")
			opt = checkMenuOption(input("\n#: "))
			if (opt):
				if (opt == "D"):
					continue
				if (opt == "S"):
					sourceFile = checkFile(input("\nInput source file path\n\n#: "))
					sourceList = [sourceFile]
				elif (opt == "F"):
					sourceDir = checkDir(input("\nInput source file path\n\n#: "))
					sourceList = bitepacker.getAudioInDir(sourceDir)
				elif (opt == "L"):
					sourceList = []
					failedListing = False
					while 1:
						newSource = input("enter path to sourcefile and hit enter, type stop to exit\n\n#: ")
						newSource = newSource.strip()
						if (newSource == "stop"):
							break
						else:
							if (not checkFile(newSource)):
								failedListing = True
								consoleBanner = "ERROR: while making a list of files - no such file!"
							else:
								sourceList.append(newSource)
					if (failedListing):
						badMessage = True
						continue

				else:
					badMessage = True
					consoleBanner('ERROR: no such option!')
					continue


				destDir = input("\nOutput slices to which directory? (blank for built-in)\n\n#: ")
				if (destDir == ""):
					destDir = "./bitepack/"


				packName = input("\nName of this new bitepack?\n\n#: ")
				if (packName.strip() == ""):
					badMessage = True
					consoleBanner = "ERROR: empty bitepack name!"
					continue

				destDir = join(destDir, packName)

				doubleSampling = checkMenuOption(input("\nUse double density sampling? (Y/N)\n\n#: "))
				if (doubleSampling == "Y"):
					d = True
				elif (doubleSampling == "N"):
					d = False
				else:
					consoleBanner = "ERROR: bad Y/N choice!"
					badMessage = True
					continue

				bitesLimit = False
				limitBitesPerFile = checkMenuOption(input("\nLimit bites per sample? (Y/N)\n\n#: "))
				if (limitBitesPerFile == "Y"):
					bitesLimit = input("\nHow many bites per one sample maximum?\n\n#: ")
					try:
						bitesLimit = int(bitesLimit)
					except Exception as e:
						badMessage = True
						consoleBanner = "ERROR: not an integer number provided!"
						continue
				elif (limitBitesPerFile != "N"):
					consoleBanner = "ERROR: bad Y/N choice!"
					badMessage = True
					continue

				if (not sourceList==False):
					check(bitesLimit)
					bitepacker.packList(sourceList, destDir, d, bitesLimit)
					packSettings = packmanager.register(destDir, packSettings)

			else:
				consoleBanner = "ERROR: bad option"
				badMessage = True

			continue;

		# run matchengine
		elif (opt == "M"):

			consoleClear()
			print("\n" + "SLOP: MatchEngine" + "\n")

			# get pack settings
			packPaths = packmanager.getActivePacksList(packSettings)


			# ask about stuff
			inFile = checkFile(input("\nProvide source file path (.wav supported so far)\n#: "))
			if (not inFile):
				badMessage = True
				consoleBanner = "ERROR: such source file does not exist!"
				continue

			outDir = checkDir(input("\nProvide directory for output (blank to put next to original)\n#: "))
			if (not outDir):
				outDir = dirname(inFile)

			outName = input("\nName of the output file without extension? (blank to use autonaming)\n#: ")
			if (outName == ""):
				outName = postfixFileName(inFile, "-limned")
			else:
				outName = join(outDir, (outName + ".wav"))


			# extended parameters
			chunkSize = 8
			fade = True
			normalize = False
			randomize = False



			extOption = checkMenuOption(input("\nEnter extended parameters? (Y/N)\n\n#: "))
			if (extOption == "Y"):
				# entering parameters here
				chunkOption = input("\nChunk length, integer from 2 to 64 (blank for default 8)\n\n#: ")
				try:
					chunkSize = int(chunkOption)
				except Exception as e:
					badMessage = True
					consoleBanner = "ERROR: non-integer value provided!"
					continue
				else:
					if ((chunkSize < 2) or (chunkSize > 64)):
						badMessage = True
						consoleBanner = "ERROR: chunk size exceeds borders!"
						continue

				tempOpt = checkMenuOption(input("\nEnable microfading? (Y/N)\n\n#: "))
				if (tempOpt == "N"):
					fade = False
				elif (tempOpt != "Y"):
					badMessage = True
					consoleBanner = "ERROR: bad (Y/N) choice"
					continue

				tempOpt = checkMenuOption(input("\nEnable chunk gain compensation? (Y/N)\n\n#: "))
				if (tempOpt == "Y"):
					normalize = True
				elif (tempOpt != "N"):
					badMessage = True
					consoleBanner = "ERROR: bad (Y/N) choice"
					continue

				tempOpt = checkMenuOption(input("\nEnable chunk length randomization? (Y/N)\n\n#: "))
				if (tempOpt == "Y"):
					maxDev = min(chunkSize, (64-chunkSize))-1
					print("\nEnter a value of possible positive AND negative deviation from entered chunk length")
					randomize = input("\nThe highest deviation available right now is: %r" % (maxDev))
					try:
						randomize = int(randomize)
					except Exception as e:
						badMessage = True
						consoleBanner = "ERROR: not an integer was provided!"
						continue
					else:
						if (randomize == 0):
							randomize = False
				elif (tempOpt != "N"):
					badMessage = True
					consoleBanner = "ERROR: bad (Y/N) choice"
					continue

			matchengine.createLimnedVersion(inFile, outDir, outName, packPaths, chunkSize, fade, normalize, randomize)

		# run packmanager
		elif (opt == "P"):

			consoleClear()
			print("\n" + "SLOP PackManager" + "\n")

			print("\nE: enable/disable bitepacks")
			print("A: enable all known bitepacks")
			print("R: rescan all known bitepack places")
			print("D: done")

			opt = checkMenuOption(input("\n#: "))

			if (opt):
				if (opt == "D"):
					continue

				elif (opt == "R"):
					r = packmanager.rescan(packPlaces)
					consoleClear()
					if (r):
						packSettings = r
						input("\nrescan complete, press any key to return...")
					else:
						input("\nRESCAN FAILED! no packs found. press any key to exit, and troubleshoot")
				elif(opt == "A"):
					packSettings = packmanager.settEnableAll(packSettings)

				elif (opt == "E"):
					consoleBanner = "SLOP PackManager: edit"
					while (1):
						consoleClear()
						print("\n" + consoleBanner + "\n")
						if (badMessage):
							badMessage = False
							consoleBanner = "SLOP PackManager: edit"
						print("\n-------- available bitepacks:")
						for num, packTuple in enumerate(packSettings["packsAvailable"]):
							print(num, basename(packTuple["path"]))
						print("\n-------- active bitepacks:")
						for item in packSettings["packsActive"]:
							print(item)
						print("\n--------\nenter the number of the pack to toggle, or done to exit manager")

						targetPack = input("\n#: ")
						if (targetPack == "done"):
							if (len(packSettings["packsActive"]) == 0):
								badMessage = True
								consoleBanner = "ERROR: can't have no packs active!"
								continue
							else:
								break
						try:
							targetPackNum = int(targetPack)
						except Exception as e:
							badMessage = True
							consoleBanner = ("ERROR: Not an integer number!")
							continue
						else:
							if ((targetPackNum > (len(packSettings["packsAvailable"])-1)) or (targetPackNum < 0)):
								badMessage = True
								consoleBanner = ("ERROR: No bitepack under such number!")
								continue
							targetTuple = packSettings["packsAvailable"][targetPackNum]
							packSettings = packmanager.settToggle(packSettings, targetTuple["path"])
					


				else:
					badMessage = True
					consoleBanner = "ERROR: Unexistant packmanager option"


			else:
				badMessage = True
				consoleBanner = "ERROR: Bad packmanager option"
				continue

		# show about
		elif (opt == "A"):
			consoleClear()
			if (not exists(aboutLocation)):
				badMessage = True
				consoleBanner = "ERROR: files are corrupt or missing! Please redownload SLOP!"
				continue
			fHandle = open(aboutLocation, "r")
			displayString = fHandle.read()
			fHandle.close()
			print(displayString)
			input("\n\npress any key to go back...")

		# show changelog
		elif (opt == "C"):
			consoleClear()
			if (not exists(changelogLocation)):
				badMessage = True
				consoleBanner = "ERROR: files are corrupt or missing! Please redownload SLOP!"
				continue
			fHandle = open(changelogLocation, "r")
			displayString = fHandle.read()
			fHandle.close()
			print(displayString)
			input("\n\npress any key to go back...")

		# exit
		elif (opt == "X"):
			packmanager.exportSettings(packSettings)
			consoleClear()
			break

		# non-existing option exception
		else:
			consoleBanner = ("ERROR: non-existent option")
			badMessage = True

	else:
		# bad input exception
		consoleBanner = ("ERROR: bad option")
		badMessage = True

#--------------------------------------------------------------------------------------------------------

inc/bitepacker.py

                #========                   BITE PACKER                   ========#
                #======== PART OF SLOP (SAMPLE-LIMNING OPERATORS PACKAGE) ========#
                #========              V1.2    16 MAR 2020                ========#

'''
MAIN FUNCTION: packList (variants: packFile, packFolder)
MAIN PURPOSE:  take a list (folder, just one) of files and process it for usage with SLOP.
               result will be a bunch of samples and an analysis file in a pre-defined folder.
NOTES:         important for making bite packs for the program.
'''

#--------------------------------------------------------------------------------------------------------
# imports

from os import listdir, rename, makedirs
from os.path import isfile, join, exists
import json

try:
	from inc import sampleslicer
	from inc import fftanalyzer
except Exception as e:
	import sampleslicer
	import fftanalyzer


#--------------------------------------------------------------------------------------------------------
# globals

maxBites = 192



#--------------------------------------------------------------------------------------------------------
# takes [directory] path, returns a list

def getAudioInDir(directory):
	supportedFormats = ['wav']
	dirContents = listdir(directory)
	output = []
	for file in dirContents:
		 if isfile(join(directory, file)):
		 	if (file[-3:] in supportedFormats):
		 		output.append(join(directory, file))
	return output


#--------------------------------------------------------------------------------------------------------
# takes in an integer and gives back a string fit to three-digit format

def fillZeroes(arg):
	out = str(arg)
	while (len(out) < 3):
		out = "0" + out
	return out


#--------------------------------------------------------------------------------------------------------
# rename all audiosamples in a folder to format PREFIXnnn, where PREFIX is any str and nnn is a 3-sym string
def renameSlices(directory, prefix="sample", fileFormat="wav"):
	fileList = getAudioInDir(directory)
	global maxBites
	counter = 0
	for file in fileList:
		if (counter > maxBites):
			return
		rename(file, join(directory, (prefix + fillZeroes(counter) + "." + fileFormat)))
		counter += 1
	return


#--------------------------------------------------------------------------------------------------------
# takes in a list of paths to wavesamples, slices them to some directory.

def sliceSampleList(listToSlice, destDir, double=True, maxSlices=False):

	global maxBites
	if (not maxSlices):
		maxSlices = maxBites

	maxBitesPerSample = min(maxBites, maxSlices)

	count = 0
	totalBites = 0


	for sample in listToSlice:

		if (totalBites == maxBites): # if we already made maxBites bites - exit to renaming them
			break
		else:
			totalBites += sampleslicer.processSample(sample, destDir, double, (str(count) + "_"),  min((maxBites - totalBites), maxBitesPerSample))
		count += 1
	renameSlices(destDir)
	return


#--------------------------------------------------------------------------------------------------------
# takes in a directory name, slices them to some directory.

def sliceSampleDir(sourceDir, destDir, double=True, maxSlices=False):
	toSlice = getAudioInDir(sourceDir)
	sliceSampleList(toSlice, destDir, double, maxSlices)
	return


#--------------------------------------------------------------------------------------------------------
# creates a list of lists of lists (file: chunk: fft values) of FFT analysis data. oh god...
def getAnalysisData(directory):
	fileList = getAudioInDir(directory)
	fftList = []
	for file in fileList:
		fftList.append(fftanalyzer.analyze(file))
	return fftList


#--------------------------------------------------------------------------------------------------------
# dumps the output of getAnalysisData to console. Don't ask why.

def dumpAnalysisDataToConsole(fftList):
	for i, fftFile in enumerate(fftList):
		print("FILE: " + str(i))
		for j, fftChunk in enumerate(fftFile):
			print("    CHUNK:" + str(j))
			for fftValue in fftChunk:
				print("            " + str(fftValue))

#--------------------------------------------------------------------------------------------------------
# returns a nicely indented JSON of the scary list of lists of lists getAnalysisData() makes.

def analysisDataToJson(adata):
	jstring = json.dumps(adata, sort_keys=True, indent=4, separators=(',', ': '))
	return jstring


#--------------------------------------------------------------------------------------------------------
# target directory gets analized (haha), and  a file is put into it. def name is sampeldata.fft. no toch

def analyzeBiteDir(directory, analysisFileName="sampledata.fft"):
	dat = getAnalysisData(directory)
	fileHandle = open(join(directory, analysisFileName), "w")
	fileHandle.write(analysisDataToJson(dat))
	return


#--------------------------------------------------------------------------------------------------------
# main function that unites it all. Processes a list of files into a folder with bites and analysis data

def packList(listToSlice, destDir, double=True, maxSlices=False):
	if (not maxSlices):
		global maxBites
		maxSlices = maxBites

	if not exists(destDir):
		makedirs(destDir)
	sliceSampleList(listToSlice, destDir, double, maxSlices)
	analyzeBiteDir(destDir)
	return


#--------------------------------------------------------------------------------------------------------
# Wrapper for packList. Processes a folder of files into a folder with bites and analysis data

def packFolder(sourceDir, destDir, double=True, maxSlices=False):
	if (not maxSlices):
		global maxBites
		maxSlices = maxBites
	fileList = getAudioInDir(sourceDir)
	packList(fileList, destDir, double)
	return


#--------------------------------------------------------------------------------------------------------
# Wrapper for packList. Processes a single file into a folder with bites and analysis data

def packFile(sourceFile, destDir, double=True, maxSlices=False):
	if (not maxSlices):
		global maxBites
		maxSlices = maxBites
	fileList = [sourceFile]
	packList(fileList, destDir, double)
	return


#--------------------------------------------------------------------------------------------------------


inc/fftanalyzer.py

                #========                  FFT ANALYZER                   ========#
                #======== PART OF SLOP (SAMPLE-LIMNING OPERATORS PACKAGE) ========#
                #========              V1.1    16 MAR 2020                ========#

'''
MAIN FUNCTION: analyze
MAIN PURPOSE:  take in a path to a *.wav file, and return a special list:
               it would contain lists of an FFT analysis of every N milliseconds of the audio
               default block size N is 2 milliseconds. So the sound better be of a multiple of 2
               milliseconds length. Unsuitable for a random sound, but it only deals with samples
               generated with the sampleslicer.py
NOTES:         this one is used in bitepacker for analysing a folder with freshly generated samples.
'''

#--------------------------------------------------------------------------------------------------------
# imports

import pydub
from pydub import AudioSegment
import matplotlib.pyplot as plt
from scipy.io import wavfile
from scipy.fftpack import fft,fftfreq
import numpy as np


#--------------------------------------------------------------------------------------------------------
# returns a mean of a list filled with floats (pwease don't put other stuff in thewe....)

def meanOfList(arg):
	return sum(arg) / len(arg) 


#--------------------------------------------------------------------------------------------------------
# get a list of amplitude describing lists, one per block, for a sample located at [sampleName]

def analyze(sampleName, blockSize=2):
	sample = AudioSegment.from_file(sampleName)
	sampleAmplitudes = []
	for i, chunk in enumerate(sample[::blockSize]):
		# if a chunk is shorter than blockSize, append some silence to it
		if (len(chunk) < blockSize):
			padLength = blockSize - len(chunk)
			silencePad = AudioSegment.silent(frame_rate=44100, duration=padLength)
			chunk = chunk.append(silencePad, crossfade=0)

		chunkAmplitudes = analyzeChunk(chunk)	
		sampleAmplitudes.append(chunkAmplitudes)
	return sampleAmplitudes


#--------------------------------------------------------------------------------------------------------
# get results of FFT analysis of a pydub AudioSegment. default sampleRanges list gen'd with rangegenerator.py

def analyzeChunk(targetSample, sampleRate=44100, sampleRanges=[0, 18, 39, 65, 95, 132, 175, 227, 289, 363, 452, 558, 685, 837, 1019, 1237, 1498, 1812, 2188, 2639, 3180, 3829, 4607, 5541, 6662, 8006, 9619, 11555, 13877, 16664, 20008, 24021]):
	
	data = targetSample.get_array_of_samples()
	data = np.array(data)
	samples = int(targetSample.frame_count())
	
	datafft = fft(data) # perform an fft
	fftabs = abs(datafft) # get its real part
	fftMax = max(fftabs) # get the maximum value of it
	freqs = fftfreq(samples,1/(sampleRate)) # get frequencies of fft

	amplitudeList = [] # list to which we put all amplitudes within a range and then average them
	meanAmplitudes = [] # list that we return, 12 averaged amplitude values end up here
	currentRange = 0 # number of the left range barrier we currently at

	# now process by every frequency that we have in the FFT result
	for count, freq in enumerate(freqs):
		# break out the moment we see a negative frequency
		if (freq < 0):
			break
		# get the amplitude that corresponds to this freq, normalize it to [0,1] range
		if (fftMax != 0):
			amplitude = fftabs[count]/fftMax
		else:
			amplitude = 0

		# if we reached a value from the next range
		if (freq > sampleRanges[currentRange+1]):
			currentRange += 1 # move to the next range
			meanAmplitudes.append(meanOfList(amplitudeList)) # average current amp. list, append it	
			amplitudeList = [] # then clear it for the next range
			if (currentRange == len(sampleRanges)-1): # if we crossed the last range right barrier - break
				break

		# put the amplitude to the current list of amplitudes for ongoing range

		amplitudeList.append(amplitude)
	return meanAmplitudes
		

#--------------------------------------------------------------------------------------------------------


inc/packmanager.py

                #========                SLOP PACK MANAGER                ========#
                #======== PART OF SLOP (SAMPLE-LIMNING OPERATORS PACKAGE) ========#
                #========              V1.0    16 MAR 2020                ========#

'''
MAIN FUNCTION: none
MAIN PURPOSE:  Manage the packs.setting file and the settings dictionary object it represents
               it controls what bitepacks the progam sees and uses 
NOTES:         -
'''

#--------------------------------------------------------------------------------------------------------
# imports

from os import listdir
from os.path import exists, isfile, join, basename
import json

try:
	from inc.bitepacker import getAudioInDir
	import inc.fftanalyzer
except Exception as e:
	from bitepacker import getAudioInDir
	import fftanalyzer


#--------------------------------------------------------------------------------------------------------
# globals

defaultAnalysisFileName = "sampledata.fft"


#--------------------------------------------------------------------------------------------------------
# returns a list of bitepack paths inside target directory. False if none found.

def listPacksInDir(dirName):
	# if no target dir return bad
	if (not exists(dirName)):
		return False

	# hook up global analysis file name setting
	global defaultAnalysisFileName

	# get all files, init output list
	dirContents = listdir(dirName)
	dirPacks = []

	# iterate, put only folders which have [defaultAnalysisFileName] file in them to out list
	for item in dirContents:
		if (not isfile(join(dirName, item))):
			if (exists(join(join(dirName, item), defaultAnalysisFileName))):
				dirPacks.append(join(dirName, item))

	# if found nothing return bad 
	if (len(dirPacks) == 0):
		return False

	return dirPacks


#--------------------------------------------------------------------------------------------------------
#

def makePackEntry(path):
	entry = {'path': path, 'sampleCount': len(getAudioInDir(path))}
	return entry


#--------------------------------------------------------------------------------------------------------
# returns a list of dicts made by makePackEntry, which describes available bitepacks

def makePackList(pathList):
	packList = []
	for packsDir in pathList:
		if (not exists(packsDir)):
			return False
		if (isfile(packsDir)):
			return False
		packsInDir = listPacksInDir(packsDir)
		if (packsInDir):
			for pack in packsInDir:
				packList.append(makePackEntry(pack))

	if (len(packList) == 0):
		return False

	return packList


#--------------------------------------------------------------------------------------------------------
# make a settings dict

def makeSettings(available, active):
	return {"packsAvailable":available, "packsActive":active}

#--------------------------------------------------------------------------------------------------------
# read settings from a file (default ./files/packs.setting)

def importSettings(inFile="./bitepack/packs.setting"):
	if (not exists(inFile)):
		return False
	fHandle = open(inFile, "r")
	jsonString = fHandle.read()
	fHandle.close()
	outSettings = json.loads(jsonString) 
	return outSettings

#--------------------------------------------------------------------------------------------------------
# write settings to a file (default ./files/packs.setting)

def exportSettings(settings, outFile="./bitepack/packs.setting"):
	outSettings = settings
	jsonString = json.dumps(outSettings, indent=4, separators=(',', ': '))
	fHandle = open(outFile, "w")
	fHandle.write(jsonString)
	fHandle.close()
	return


#--------------------------------------------------------------------------------------------------------
# settings modifier: enables all available bitepacks in the settings list

def settEnableAll(set):
	settings = set;
	settings["packsActive"] = []
	for i in settings["packsAvailable"]:
		settings["packsActive"].append(basename(i["path"]))
	return settings


#--------------------------------------------------------------------------------------------------------
# settings modifier: enables a pack by path

def settToggle(set, path):
	settings = set
	for i in settings["packsAvailable"]:
		# if there is such a pack
		if (i["path"] == path):
			if (basename(path) in settings["packsActive"]):
				settings["packsActive"].remove(basename(path))
			else:
				settings["packsActive"].append(basename(path))
			return settings
	return False


#--------------------------------------------------------------------------------------------------------
# returns a list with paths to bitepacks that are enable in the settings dict

def getActivePacksList(settings):
	outList = []
	for packTuple in settings["packsAvailable"]:
		if (basename(packTuple["path"]) in settings["packsActive"]):
			outList.append(packTuple["path"])
	return outList


#--------------------------------------------------------------------------------------------------------
# rescan palces for bitepacks and rebuild a setting

def rescan(places):
	# places are folders where bitepack folders MIGHT be. we rescan every of them
	packData = makePackList(places)

	if (len(packData) == 0):
		return False
	settings = makeSettings(packData, [])
	settings = settEnableAll(settings)
	return settings

#--------------------------------------------------------------------------------------------------------
# register a new pack into the running setting

def register(path, set):
	packData = makePackEntry(path)
	newSet = set
	newSet["packsAvailable"].append(packData)
	newSet["packsActive"].append(basename(path))
	return newSet


inc/matchengine.py

                #========                   MATCH ENGINE                  ========#
                #======== PART OF SLOP (SAMPLE-LIMNING OPERATORS PACKAGE) ========#
                #========              V1.0    16 MAR 2020                ========#

'''
MAIN FUNCTION: matchSample()
MAIN PURPOSE:  build a limned version of an input sample out of available packs
NOTES:         just a small additional ware. Might include it as a small utility in the CLI or GUI ver.
'''

#--------------------------------------------------------------------------------------------------------
# imports

import json
import pydub
from pydub import AudioSegment
from os.path import isfile, join
from math import ceil
from random import randrange

try:
	from inc import fftanalyzer
	from inc.bitepacker import fillZeroes
except:
	import fftanalyzer
	from bitepacker import fillZeroes


#--------------------------------------------------------------------------------------------------------
# make a limned version of inFile and render it to outDirectory under name outName.wav

def createLimnedVersion(inFile, outDir, outName, bitePackList, chunkSize=4, fade=False, normalize=False, randomize=False ):
	sampleFFTData = getSampleFFT(inFile)
	newSample = matchSample(sampleFFTData, bitePackList, chunkSize, fade, normalize, randomize)
	outPath = join(outDir, outName)
	file_handle = newSample.export(outPath, format="wav")


#--------------------------------------------------------------------------------------------------------
# takes in a path to input file, returns its FFT data

def getSampleFFT(sampleName, blockSize=2):
	dat = fftanalyzer.analyze(sampleName, blockSize)
	return dat


#--------------------------------------------------------------------------------------------------------
# takes in two FFT Amplitude lists, 

def getAmpListDistance(l1, l2):
	if (len(l1) != len(l2)):
		raise ValueError("chunk numbers differ in analysis file and current running routine!!!")
	distance = 0
	for num, val1 in enumerate(l1):
		distance += abs(l1[num] - l2[num])
	return distance


#--------------------------------------------------------------------------------------------------------
# read a bitepack analysis file and convert it to list of lists of lists (gosh, i am getting TIRED...)

def jsonToBitePackAnalysisList(inFile):
	fHandle = open(inFile, "r")
	jsonString = fHandle.read()
	fHandle.close()
	analysis = json.loads(jsonString)
	return analysis


#--------------------------------------------------------------------------------------------------------
# get chunk of amplitude lists of a bite number [sampleNum] from a pack described by [fftList] list

def getBiteChunk(sampleNum, fftList, size):
	# if we requesting a negative or a nonexisting sample - tell to go yiff self
	if ((sampleNum < 0) or (sampleNum > (len(fftList)-1))):
		raise ValueError("Sample not exists!")
	out = fftList[sampleNum]
	return out[:size]


#--------------------------------------------------------------------------------------------------------

def matchSample(sampleFFTData, bitePackList, givenChunkSize=4, fade=False, norm=False, rand=False, blockSize=2, dataFileName="sampledata.fft"):

	# had to move the original value into a variable, more or less the CLI calls it 'chunk length', so...
	# chunk size is value used DURING APPENDAGE, chunk length is the original value passed on when called
	chunkSize = givenChunkSize

 	# fade-related values calculation
	fadeInChunks = ceil(chunkSize/4)

	if (fade):
		sampleSize = blockSize*(chunkSize+fadeInChunks)
	else:
		sampleSize = blockSize*chunkSize

	# bitePackList = list of bitepack directories.
	bpDirAndData = []

	for bitePackDir in bitePackList:
		fftFile = join(bitePackDir, dataFileName)
		data = jsonToBitePackAnalysisList(fftFile)
		tuple = [bitePackDir, data]
		bpDirAndData.append(tuple)

	outSample = AudioSegment.silent(duration=0, frame_rate=44100)
	outSample = outSample.empty()

	count = 0
	cfade = 0

	while (count < len(sampleFFTData)):

		if (count+chunkSize < len(sampleFFTData)):
			currentSampleChunk = sampleFFTData[count:count+(chunkSize-1)]
		else:
			currentSampleChunk = sampleFFTData[count:]
		
		goodSampleTuples = []
		for tuple in bpDirAndData:
			goodSampleTuples.append(matchChunk(currentSampleChunk, tuple[1], len(currentSampleChunk)))

		# finally find the BEST bite to represent current chunk of the in sample
		minDistance = 1024
		bestBitePack = 0
		for num, val in enumerate(goodSampleTuples):
			if (val[1] < minDistance):
				minDistance = val[1]
				bestBitePack = num

		appendSampleName = "sample" + fillZeroes(goodSampleTuples[bestBitePack][0]) + ".wav"
		appendSamplePath = join(bpDirAndData[bestBitePack][0], appendSampleName)


		appendSample = AudioSegment.from_file(appendSamplePath, format="wav")


		if ((fade) and (count>0)):
			cfade = blockSize*fadeInChunks

		appendSample = appendSample[:sampleSize]

		if (norm):
			appendSample = normalizeAudioSegment(appendSample)

		outSample = outSample.append(appendSample, crossfade=cfade)

		count += chunkSize

		# post-appendage random chunk and fade length recalculation if needed:
		if (rand):
			chunkSize = givenChunkSize + randrange(-rand, +rand)
			fadeInChunks = ceil(givenChunkSize/4)

	return outSample

#--------------------------------------------------------------------------------------------------------
#

def matchChunk(sampleChunkData, fftList, chunkSize):


	distance = 0
	minDistance = 1024
	bestSampleNum = 0

	for num, bite in enumerate(fftList):
		biteChunkData = getBiteChunk(num, fftList, chunkSize)
		distance = compareChunks(sampleChunkData, biteChunkData)
		if (distance < minDistance):
			minDistance = distance
			bestSampleNum = num

	return [bestSampleNum, minDistance]

#--------------------------------------------------------------------------------------------------------
#

def compareChunks(sampleChunk, biteChunk):

	if (len(sampleChunk) != len(biteChunk)):
		raise ValueError("Awful garbage has happened, blame Astro!")
	overallDistance = 0
	for num, val in enumerate(sampleChunk):
		overallDistance += getAmpListDistance(sampleChunk[num], biteChunk[num])
	return overallDistance


#--------------------------------------------------------------------------------------------------------
# return a volume-normalized copy of the sample

def normalizeAudioSegment(aseg):
	return aseg.apply_gain(-aseg.max_dBFS)




#--------------------------------------------------------------------------------------------------------
